torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   abort(abortReason);
 }
 
+void dump_debugging_info() {
+  LOG(ERROR) << "Dumped debugging information: " << dump_nccl_trace();
+}
+
+void ProcessGroupNCCL::ncclCommsMonitor() {
+  int hearBeatCounter = -1;
+  while (true) {
+    auto heartbeat = heartbeat_.load();
+    if (heartbeat > hearBeatCounter) {
+      hearBeatCounter = heartbeat;
+      // Sleep for enough time so that heartBeat will at least increase.
+      std::this_thread::sleep_for(
+          std::chrono::milliseconds(kWatchdogThreadSleepMillis * 3));
+    } else {
+      // Timeout case and we will first dump the flight recorder
+      // to std::out, and we use a different detached thread to ensure
```

Comment:
i might structure this differently- 
```
else {
  // we clearly exit the loop on first error
 break;
}

// then below we do the 'epilogue' behavior for dumping and crashing
```

also, maybe skip the sideThread for now.  Is dump actually unreliable?  If so we might want a side thread.  (later, if we're dumping to manifold for example).

But in this case, we can probably just call dump, and then ensure we exit/exception

Reply:
I see. Originally @kwen2501 suggested to use a side thread to keep this thread as clean as possible. For now since we are not doing very complicated thing, I am ok with either way. I am not sure if dumping is unreliable or not. But it's like an extra protection?

Reply:
Let's only add the side thread if we're sure about it.  It otheriwse adds complexity that might get in the way later.

One example is that if we were to launch some RPC calls here, the RPC calls may use their own 'side-thread' so that the calls can return immediately.  In that case we probably don't want our own side-thread.  But if we decide instead to call risky/expensive functions inline we probably should use a side-thread.

However, if we do use a side thread, then we have to consider how long to let it work before we proceed to exit.

Is our plan still that the watchdog stuck will kill the process?  (if so, raising an exception is not good enough, becuase exceptions can call destructors which can then acquire locks and hang).  So maybe we need to call std::abort() at the end of this function to be safe.

Then, I see 2 options
(a) we call dump_debug_info() inside this thread, and then call std::abort().  We must ensure dump_debug_info is a safe function to call.  (if all it does is print some stuff thats in memory, it should not be hard to analyze).  If it needs to check on any more work objects or make additional cuda calls, this would not be good.
(b) use a side thread to call dump_debug_info.   but, then we need to sleep this thread for some time to let the side-thread work, before we finally call abort.

Reply:
AFAIK, by reading through flight recorder's PR, there seems to be no CUDA call, so it should be ok that we call it here directly. And I see your point, because this is the really last resort, we need to abort the process anyway rather than throwing exceptions.. But that makes unit test hard, there is no way we can do try catch..